// node-red-contrib-tuyapi-cloud/build/index.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const TuyaCloud = require("@tuyapi/cloud");
// import { get } from 'lodash';
module.exports = (RED) => {
    function configuration(config) {
        RED.nodes.createNode(this, config);
        this.name = config.name;
        this.key = config.key;
        this.secret = config.secret;
        this.secret2 = config.secret2;
        this.certSign = config.certSign;
        this.region = config.region;
        this.apiEtVersion = config.apiEtVersion;
        this.email = config.email;
        this.password = config.password;
        this.mapping = config.mapping;
    }
    function request(config) {
        RED.nodes.createNode(this, config);
        const node = this;
        const conf = RED.nodes.getNode(config.config);
        // console.dir({ conf: Object.keys(conf) }, {color: true});
        node.status({ fill: 'yellow', shape: 'dot', text: 'connecting' });
        let api;
        try {
            api = new TuyaCloud({
                key: conf.key,
                secret: conf.secret,
                secret2: conf.secret2,
                certSign: conf.certSign,
                apiEtVersion: conf.apiEtVersion,
                region: conf.region,
            });
        }
        catch (e) {
            node.status({ fill: 'red', shape: 'ring', text: e.message });
            throw e;
        }
        let sid;
        api.loginEx({ email: conf.email, password: conf.password })
            .then((sessionId) => {
            node.status({ fill: 'green', shape: 'dot', text: 'connected' });
            sid = sessionId;
            node.log(`Session Id: ${JSON.stringify(sid)}`);
        }).catch((e) => {
            node.status({ fill: 'red', shape: 'ring', text: `Unable to connect: ${e.message}` });
            node.error(`Error Logging in: ${JSON.stringify(e)}`);
        });
        node.on('input', (msg) => __awaiter(this, void 0, void 0, function* () {
            const { action, group: gid, requireSID, data } = msg.payload;
            msg.payload = yield api.request({ action, gid, data, requireSID }).catch((e) => {
                node.error(`Error Requesting: ${JSON.stringify(e)}`);
            });
            return node.send(msg);
        }));
    }
    RED.nodes.registerType('tuyapi-cloud-request', request);
    RED.nodes.registerType('tuyapi-cloud-configuration', configuration);
};
